src/xmipp/libraries/reconstruction/tomo_filter_coordinates.cpp

```diff
@@ -0,0 +1,248 @@
+		{
+			std::cout << "WARNNING: Coordinate at (x=" << coord3D.x<< ", y=" << coord3D.y << ", z=" << coord3D.z << ") erased due to its out of the mask." << std::endl;
+			inputCoords.erase(inputCoords.begin()+i);
+			i--;
+		}
+
+		else if(DIRECT_A3D_ELEM(inputVolume, coord3D.z, coord3D.y, coord3D.x) == 0)
+		{
+			#ifdef VERBOSE_OUTPUT
+			std::cout << "Coordinate erased with value " << DIRECT_A3D_ELEM(inputVolume, coord3D.z, coord3D.y, coord3D.x) << " at (x=" << coord3D.x << ", y=" << coord3D.y << ", z=" << coord3D.z << ")" << std::endl;
+			#endif 
+			inputCoords.erase(inputCoords.begin()+i);
+			i--;
+		}
+		
+		#ifdef VERBOSE_OUTPUT
+		else
+		{
+			std::cout << "Coordinate saved with value " << DIRECT_A3D_ELEM(inputVolume, coord3D.z, coord3D.y, coord3D.x) << " at (x=" << coord3D.x << ", y=" << coord3D.y << ", z=" << coord3D.z << ")" << std::endl;
+		}
+		#endif 
+	}
+}
+
+
+void ProgTomoFilterCoordinates::takeCoordinateFromTomo(MultidimArray<double> &tom)
+{
+	MetaDataVec scoredMd;
+	MDRowVec row;
+
+	for (size_t i = 0; i < inputCoords.size(); i++)
+	{
+		Point3D coor = inputCoords[i];
+
+		if (((coor.z - radius) < 0) || ((coor.z + radius) > (zDim-1)) || ((coor.y - radius) < 0) || ((coor.y + radius) > (yDim-1)) || ((coor.x - radius) < 0) || ((coor.x + radius) > (xDim-1)))
+		{
+			std::cout << "WARNNING: Coordinate at (x=" << coor.x<< ", y=" << coor.y << ", z=" << coor.z << ") erased due to its out of the mask." << std::endl;
+			continue;
+		}
+				
+		double meanCoor = 0;
+		double meanCoor2 = 0;
+		double stdCoor = 0;
+		// double medianCoor = 0;
+		// double madCoor = 0;
+		double value = 0;
+		size_t Nelems = 0;
+		
+
+		for (int i = -radius; i < radius; i++)
+		{
+			size_t i2 = i*i;
+			for (int j = -radius; j < radius; j++)
+			{
+				size_t j2i2 = i2 + j*j;
+				for (int k = -radius; k < radius; k++)
+				{
+					size_t r2 = j2i2 + k*k;
+					
+					if (r2 <= radius)
+					{
+						double value;
+						value = DIRECT_A3D_ELEM(tom, coor.z + k, coor.y + i, coor.x + j);
+						meanCoor += value;
+						Nelems++;
+						meanCoor2 += value*value;
+					}
+				}
+			}
+		}
+
+		meanCoor /= Nelems;
+		stdCoor = sqrt(meanCoor2/Nelems - meanCoor*meanCoor);
+		
+		row.setValue(MDL_XCOOR, coor.x);
+		row.setValue(MDL_YCOOR, coor.y);
+		row.setValue(MDL_ZCOOR, coor.z);
+		row.setValue(MDL_AVG, meanCoor);
+		row.setValue(MDL_STDDEV, stdCoor);
+		// row.setValue(MDL_VOLUME_SCORE1, medianCoor);
+		// row.setValue(MDL_VOLUME_SCORE2, madCoor);
+		scoredMd.addRow(row);
+	}
+	scoredMd.write(fnOut);
+}
+
+// --------------------------- I/O functions ----------------------------
+
+void ProgTomoFilterCoordinates::readInputCoordinates()
+{
+	MetaDataVec inCoordMd;
+	inCoordMd.read(fnInCoord);
+
+	size_t objId;
+	Point3D<int> coordinate3D;
+
+	for(size_t objId : inCoordMd.ids())
+	{
+		inCoordMd.getValue(MDL_XCOOR, coordinate3D.x, objId);
```

Comment:
are you aware that `getValue` returns bool, stating whether `coordinate3d` has been updated? You probably want to use `getValueOrDefault` or explicitly check that the metadata has these labels

Reply:
your program will not fail. If the value is not found in the metadata, you will get some garbage value. To avoid checking the value in each iteration, you can query the metadata before you enter the loop, e.g. in the `run()` method. Look for sth like `inCoordMd.hasLabel()`.

Reply:
MetadataVec class has not function hasLabel. 
error: 'class MetaDataVec' has no member named 'hasLabel'

Also, getValue() method throws an error if value is not found 

    T& getValue(const MDLabel label, size_t id) {
        MDObject mdValueOut(label);
        if (!getValue(mdValueOut, id))
            throw ObjectDoesNotExist();
        return mdValueOut.getValue2(T());
    }